File: backend/src/main/java/com/example/project_CS261/service/FavoriteService.java
package com.example.project_CS261.service;

import com.example.project_CS261.model.Event;
import com.example.project_CS261.model.Favorite;
import com.example.project_CS261.model.NotificationQueue;
import com.example.project_CS261.repository.EventRepository;
import com.example.project_CS261.repository.FavoriteRepository;
import com.example.project_CS261.repository.NotificationQueueRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@Transactional
public class FavoriteService {

    private static final Logger logger = LoggerFactory.getLogger(FavoriteService.class);

    private final FavoriteRepository favoriteRepository;
    private final NotificationQueueRepository notificationQueueRepository;
    private final EventRepository eventRepository;

    public FavoriteService(FavoriteRepository favoriteRepository, 
                          NotificationQueueRepository notificationQueueRepository, 
                          EventRepository eventRepository) {
        this.favoriteRepository = favoriteRepository;
        this.notificationQueueRepository = notificationQueueRepository;
        this.eventRepository = eventRepository;
    }

    /**
     * Add event to user's favorites
     */
    public Favorite addFavorite(Long userId, Long eventId) {
        logger.info("=== Adding favorite ===");
        logger.info("userId: {}", userId);
        logger.info("eventId: {}", eventId);
        
        // Validate inputs
        if (userId == null) {
            logger.error("userId is NULL!");
            throw new IllegalArgumentException("User ID cannot be null");
        }
        
        if (eventId == null) {
            logger.error("eventId is NULL!");
            throw new IllegalArgumentException("Event ID cannot be null");
        }
        
        // Check if already favorited
        if (favoriteRepository.existsByUserIdAndEventId(userId, eventId)) {
            logger.warn("Event {} already favorited by user {}", eventId, userId);
            throw new RuntimeException("Event already favorited");
        }

        // Create favorite
        Favorite favorite = new Favorite();
        favorite.setUserId(userId);
        favorite.setEventId(eventId);

        // Get event details for notification
        Event event = eventRepository.findById(eventId)
                .orElseThrow(() -> new RuntimeException("Event not found with id: " + eventId));

        logger.info("Event found: {}", event.getTitle());
        logger.info("Event start time: {}", event.getStartTime());

        // Create notification queue entry
        NotificationQueue nq = new NotificationQueue();
        nq.setUserId(userId);
        nq.setEventId(eventId);
        nq.setSendAt(event.getStartTime().minusDays(1));
        nq.setStatus("PENDING");
        
        logger.info("NotificationQueue before save:");
        logger.info("  userId: {}", nq.getUserId());
        logger.info("  eventId: {}", nq.getEventId());
        logger.info("  sendAt: {}", nq.getSendAt());
        logger.info("  status: {}", nq.getStatus());
        
        notificationQueueRepository.save(nq);
        logger.info("NotificationQueue saved successfully");

        Favorite savedFavorite = favoriteRepository.save(favorite);
        logger.info("Favorite added successfully for user {} and event {}", userId, eventId);
        
        return savedFavorite;
    }

    /**
     * Get all favorites for a user
     */
    public List<Favorite> getFavoritesByUser(Long userId) {
        logger.info("Fetching favorites for user: {}", userId);
        return favoriteRepository.findByUserId(userId);
    }

    /**
     * Remove event from user's favorites
     */
    public void removeFavorite(Long userId, Long eventId) {
        logger.info("Removing favorite: userId={}, eventId={}", userId, eventId);
        
        // Remove from notification queue
        notificationQueueRepository.deleteByUserIdAndEventId(userId, eventId);
        
        // Remove favorite
        favoriteRepository.deleteByUserIdAndEventId(userId, eventId);
        
        logger.info("Favorite removed successfully for user {} and event {}", userId, eventId);
    }

    /**
     * Check if event is favorited by user
     */
    public boolean isFavorited(Long userId, Long eventId) {
        return favoriteRepository.existsByUserIdAndEventId(userId, eventId);
    }
}

File: frontend/src/services/favoriteService.js
import axiosInstance from '../lib/axiosInstance';

/**
 * เพิ่ม Event ลง Favorites
 * @param {number} eventId - ID ของกิจกรรม
 * @param {number} userId - ID ผู้ใช้
 * @returns {Promise<Object>}
 */
export const addFavorite = async (eventId, userId) => {
  try {
    if (!userId) throw new Error('User ID is required');
    const response = await axiosInstance.post('/favorites', {
      userId,
      eventId,
    });
    return response.data;
  } catch (error) {
    console.error('Error adding favorite:', error);
    throw new Error(error.message || 'ไม่สามารถบันทึกกิจกรรมได้');
  }
};

/**
 * ลบ Event ออกจาก Favorites
 * @param {number} eventId - ID ของกิจกรรม
 * @param {number} userId - ID ผู้ใช้
 * @returns {Promise<void>}
 */
export const removeFavorite = async (eventId, userId) => {
  try {
    if (!userId) throw new Error('User ID is required');
    await axiosInstance.delete('/favorites', {
      data: { userId, eventId },
    });
  } catch (error) {
    console.error('Error removing favorite:', error);
    throw new Error(error.message || 'ไม่สามารถลบกิจกรรมออกจากรายการโปรดได้');
  }
};

/**
 * ดึงรายการ Favorites ของ User ที่ login อยู่
 * @returns {Promise<Array>}
 */
export const getFavoritesByUser = async () => {
  try {
    const response = await axiosInstance.get('/favorites/user');
    return response.data;
  } catch (error) {
    console.error('Error fetching favorites:', error);
    throw new Error(error.message || 'ไม่สามารถโหลดรายการโปรดได้');
  }
};

/**
 * ตรวจสอบว่า event ถูก favorite หรือยัง
 * @param {number} eventId - ID ของกิจกรรม
 * @returns {Promise<boolean>}
 */
export const checkIsFavorited = async (eventId) => {
  try {
    const response = await axiosInstance.get(`/favorites/check/${eventId}`);
    return response.data;
  } catch (error) {
    console.error('Error checking favorite status:', error);
    return false;
  }
};

/**
 * ดึงจำนวน favorites ของ event
 * @param {number} eventId - ID ของกิจกรรม
 * @returns {Promise<number>}
 */
export const getFavoriteCount = async (eventId) => {
  try {
    const response = await axiosInstance.get(`/favorites/event/${eventId}/count`);
    return response.data;
  } catch (error) {
    console.error('Error fetching favorite count:', error);
    return 0;
  }
};

/**
 * Toggle Favorite (เพิ่มหรือลบ)
 * @param {number} eventId
 * @param {boolean} isFavorited - สถานะปัจจุบัน
 * @returns {Promise<Object|void>}
 */
export const toggleFavorite = async (eventId, isFavorited, userId) => {
  if (isFavorited) {
    // ถ้า favorite อยู่แล้ว ให้ลบออก
    return await removeFavorite(eventId, userId);
  } else {
    // ถ้ายังไม่ favorite ให้เพิ่ม
    return await addFavorite(eventId, userId);
  }
};
